// starknet/types/api/openrpc.js
"use strict";
/**
 * Starknet RPC version 0.2.0
 *
 * StarkNet Node API 0.45.0 - rpc 0.2.1
 * StarkNet Node Write API 0.3.0 - rpc 0.2.1
 * StarkNet Trace API 0.4.0 - rpc 0.2.1
 *
 * TypeScript Representation of OpenRpc protocol types
 */
Object.defineProperty(exports, "__esModule", { value: true });
var BLOCK_TAG;
(function (BLOCK_TAG) {
    BLOCK_TAG[BLOCK_TAG["latest"] = 0] = "latest";
    BLOCK_TAG[BLOCK_TAG["pending"] = 1] = "pending";
})(BLOCK_TAG || (BLOCK_TAG = {}));
var L1_HANDLER;
(function (L1_HANDLER) {
    L1_HANDLER[L1_HANDLER["L1_HANDLER"] = 0] = "L1_HANDLER";
})(L1_HANDLER || (L1_HANDLER = {}));
var STRUCT_ABI_TYPE;
(function (STRUCT_ABI_TYPE) {
    STRUCT_ABI_TYPE[STRUCT_ABI_TYPE["struct"] = 0] = "struct";
})(STRUCT_ABI_TYPE || (STRUCT_ABI_TYPE = {}));
var EVENT_ABI_TYPE;
(function (EVENT_ABI_TYPE) {
    EVENT_ABI_TYPE[EVENT_ABI_TYPE["event"] = 0] = "event";
})(EVENT_ABI_TYPE || (EVENT_ABI_TYPE = {}));
var FUNCTION_ABI_TYPE;
(function (FUNCTION_ABI_TYPE) {
    FUNCTION_ABI_TYPE[FUNCTION_ABI_TYPE["function"] = 0] = "function";
    FUNCTION_ABI_TYPE[FUNCTION_ABI_TYPE["l1_handler"] = 1] = "l1_handler";
    FUNCTION_ABI_TYPE[FUNCTION_ABI_TYPE["constructor"] = 2] = "constructor";
})(FUNCTION_ABI_TYPE || (FUNCTION_ABI_TYPE = {}));
var CALL_TYPE;
(function (CALL_TYPE) {
    CALL_TYPE[CALL_TYPE["DELEGATE"] = 0] = "DELEGATE";
    CALL_TYPE[CALL_TYPE["CALL"] = 1] = "CALL";
})(CALL_TYPE || (CALL_TYPE = {}));
var ENTRY_POINT_TYPE;
(function (ENTRY_POINT_TYPE) {
    ENTRY_POINT_TYPE[ENTRY_POINT_TYPE["EXTERNAL"] = 0] = "EXTERNAL";
    ENTRY_POINT_TYPE[ENTRY_POINT_TYPE["L1_HANDLER"] = 1] = "L1_HANDLER";
    ENTRY_POINT_TYPE[ENTRY_POINT_TYPE["CONSTRUCTOR"] = 2] = "CONSTRUCTOR";
})(ENTRY_POINT_TYPE || (ENTRY_POINT_TYPE = {}));
